python/ql/src/Resources/FileNotAlwaysClosedQuery.qll

```diff
@@ -0,0 +1,137 @@
+  predicate guardsExceptions(DataFlow::CfgNode raises) {
+    this.asCfgNode() = raises.asCfgNode().getAnExceptionalSuccessor().getASuccessor*()
+    or
+    // The expression is after the close call.
+    // This also covers the body of a `with` statement.
+    raises.asCfgNode() = this.asCfgNode().getASuccessor*()
+  }
+}
+
+/** A call to the `.close()` method of a file object. */
+class FileCloseCall extends FileClose {
+  FileCloseCall() { exists(DataFlow::MethodCallNode mc | mc.calls(this, "close")) }
+}
+
+/** A call to `os.close`. */
+class OsCloseCall extends FileClose {
+  OsCloseCall() { this = API::moduleImport("os").getMember("close").getACall().getArg(0) }
+}
+
+/** A `with` statement. */
+class WithStatement extends FileClose {
+  WithStatement() { this.asExpr() = any(With w).getContextExpr() }
+}
+
+/** Holds if an exception may be raised at `raises` if `file` is a file object. */
+private predicate mayRaiseWithFile(DataFlow::CfgNode file, DataFlow::CfgNode raises) {
+  // Currently just consider any method called on `file`; e.g. `file.write()`; as potentially raising an exception
+  raises.(DataFlow::MethodCallNode).getObject() = file and
+  not file instanceof FileOpen and
+  not file instanceof FileClose
+}
+
+/** Holds if data flows from `nodeFrom` to `nodeTo` in one step that also includes file wrapper classes. */
+private predicate fileLocalFlowStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+  DataFlow::localFlowStep(nodeFrom, nodeTo)
+  or
+  exists(FileWrapperCall fw | nodeFrom = fw.getWrapped() and nodeTo = fw)
+}
+
+/** Holds if data flows from `source` to `sink`, including file wrapper classes. */
+private predicate fileLocalFlow(DataFlow::Node source, DataFlow::Node sink) {
+  fileLocalFlowStep*(source, sink)
+}
+
+/** Holds if the file opened at `fo` is closed. */
+predicate fileIsClosed(FileOpen fo) { exists(FileClose fc | fileLocalFlow(fo, fc)) }
+
+/** Holds if the file opened at `fo` is returned to the caller. This makes the caller responsible for closing the file. */
+predicate fileIsReturned(FileOpen fo) {
+  exists(Return ret, Expr retVal |
+    (
+      retVal = ret.getValue()
+      or
+      retVal = ret.getValue().(List).getAnElt()
+      or
+      retVal = ret.getValue().(Tuple).getAnElt()
+    ) and
+    fileLocalFlow(fo, DataFlow::exprNode(retVal))
+  )
+}
+
+/** Holds if the file opened at `fo` is stored in a field. We assume that another method is then responsible for closing the file. */
+predicate fileIsStoredInField(FileOpen fo) {
+  exists(DataFlow::AttrWrite aw | fileLocalFlow(fo, aw.getValue()))
+}
+
+/** Holds if the file opened at `fo` is not closed, and is expected to be closed. */
+predicate fileNotClosed(FileOpen fo) {
+  not fileIsClosed(fo) and
+  not fileIsReturned(fo) and
+  not fileIsStoredInField(fo)
+}
+
+predicate fileMayNotBeClosedOnException(FileOpen fo, DataFlow::Node raises) {
+  fileIsClosed(fo) and
+  exists(DataFlow::CfgNode fileRaised |
```

Comment:
Maybe move the `raises` argument into this `exists` (as it doesn't seem to be used anywhere outside of this predicate)?

Reply:
At some point I had considered using `raises` in the query alert; but wasn't sure how to handle that the "open and not closed" case with no exception involved wouldn't have that information. I suppose introducing an `OptionalDataFlowNode` as an IPA type could work.

Reply:
Fair enough. We can just leave it in for now. 👍